cilium: Fix 16bit ifindex limitation

borkmann opened the pull request:
See details in commit desc.

Fixes: #16260

borkmann committed:
cilium: Fix 16bit ifindex limitation

The limitation exists mainly on old kernels where the fib lookup helper
does not populate the outgoing ifindex. Only for this case we rely on
the CT lookup stored ifindex which back then was added as a 16bit field
due to limited padding space available. Nowadays this can be lifted
after the big rework in #23884. We've seen users with high netdevice
churn run into this limitation where the agent bails out.

Apart from fixing the bleed, this can be further refined by not relying
on the asm.FnRedirectPeer helper presence but by actually doing a runtime
BPF program probe so that stable kernels can even be covered.

Fixes: #16260
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

julianwiedmann commented on bpf/lib/conntrack.h:
```diff
@@ -246,7 +246,8 @@ __ct_lookup(const void *map, struct __ctx_buff *ctx, const void *tuple,
 			ct_state->from_tunnel = entry->from_tunnel;
-			ct_state->ifindex = entry->ifindex;
+			if (!is_defined(HAVE_FIB_IFINDEX))
+				ct_state->ifindex = entry->ifindex;
```
Let's also wrap the  `ct_state->ifindex` users in `nodeport_rev_dnat_ingress_ipv*()` ? Otherwise this looks like the `ifindex` wasn't even used anymore :).

And maybe also the `ifindex` field in the `ct_state` itself, similar as for `DISABLE_LOOPBACK_LB`.

borkmann replied:
Agree, good point. ct_state is not part of map, so this is doable.

julianwiedmann reviewed (changes requested):
lgtm, thank you! Needs just one small change I think (see review below).

As follow-on, let's also add `HAVE_FIB_IFINDEX` to the relevant complexity tests to reflect the real world.

> Apart from fixing the bleed, this can be further refined by not relying
on the asm.FnRedirectPeer helper presence but by actually doing a runtime
BPF program probe so that stable kernels can even be covered.

We can further also check for `ENABLE_SKIP_FIB`.

borkmann committed:
bpf, complexity-tests: Add HAVE_FIB_IFINDEX coverage

Commit d1c362e1dd68 ("bpf: Always return target ifindex in bpf_fib_lookup")
which HAVE_FIB_IFINDEX reflects is part of is 5.10+ kernels. Add the define
to the complexity tests for 5.10 and net-next to better reflect real world.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

julianwiedmann reviewed (approved):
looks good, ty!

ti-mo reviewed (approved):
Thanks!
